addons/mail/controllers/google_translate.py

```diff
@@ -0,0 +1,49 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.http import request, route, Controller
+
+import babel
+import requests
+
+
+class GoogleTranslateController(Controller):
+    @route("/mail/message/translate", type="json", auth="user")
+    def translate(self, message_id):
+        message = request.env["mail.message"].browse(message_id)
+        message.ensure_one()
+        message._validate_access_for_current_persona("read")
```

Comment:
`ensure_one` gives no guaranteed on existence or access of message, it's just checking the size of recordset which is trivially 1 when using `browse` with one id.

And `browse` gives no guarantee either as it just builds a dummy recordset.

Either:
- combine `browse` with `exists` to verify the existence in db
- or use `search` (which is equivalent as it will make a query directly)

Then `ensure_one` will make sense.

Reply:
> "it's just checking the size of recordset"
Well, that is exactly what I was looking for because the code below should not run if the recordset is empty and it was simpler than adding another "if" and raise an error myself.

However, I noticed that the vast majority of browse() in the mail/ module are followed by exists() and I will make sure to to it as well.

Reply:
Thank you for the explanation I understand why my previous implementation did not make sense